File: Reorder List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if head.next is None:
            return

            # Find midpoint
        slow, fast = head, head.next
        while fast.next is not None:
            fast = fast.next
            slow = slow.next

            if fast.next is None:
                break
            fast = fast.next

        pt = slow.next

        # Detach midpoint
        slow.next = None

        # Reverse second half
        n = pt.next
        pt.next = None
        while n is not None:
            n2 = n.next
            n.next = pt
            pt, n = n, n2

        pt1, pt2 = head, pt
        while pt1 is not None and pt2 is not None:
            pt3 = pt2.next
            pt2.next = pt1.next
            pt1.next = pt2
            pt1, pt2 = pt1.next.next, pt3


# # Definition for singly-linked list.
# # class ListNode:
# #     def __init__(self, val=0, next=None):
# #         self.val = val
# #         self.next = next
# class Solution:
#     def reorderList(self, head: Optional[ListNode]) -> None:
#         """
#         Do not return anything, modify head in-place instead.
#         """
#         if head.next is None:
#             return
#
#         slow, fast = head, head.next
#         while fast.next is not None:
#             fast = fast.next
#             slow = slow.next
#
#             if fast.next is None:
#                 break
#             fast = fast.next
#
#         prev, pt = slow, slow.next
#         slow.next = None
#         while pt is not None:
#             next = pt.next
#             pt.next = prev
#             prev, pt = pt, next
#
#         start, end = head, prev
#         # print(head)
#         # print(end)
#         while start is not None:
#             next = start.next
#             start.next = end
#             next_end = end.next
#             end.next = next
#
#             start, end = next, next_end
#

File: Reorder List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reorderList(head *ListNode)  {
    if head.Next == nil {
        return
    }

    slow := head
    fast := head.Next
    for fast.Next != nil {
        fast = fast.Next
        slow = slow.Next

        if fast.Next == nil {
            break
        }
        fast = fast.Next
    }

    pt := slow.Next
    slow.Next = nil
    n := pt.Next
    pt.Next = nil

    for n != nil {
        n2 := n.Next
        n.Next = pt
        pt = n
        n = n2
    }

    pt1 := head
    pt2 := pt
    for pt1 != nil && pt2 != nil {
        pt3 := pt2.Next
        pt2.Next = pt1.Next
        pt1.Next = pt2
        pt1 = pt1.Next.Next
        pt2 = pt3
    }

}

